orchagent/dash/dashaclorch.cpp

```diff
@@ -0,0 +1,881 @@
+
+    if (buffer == "true" || buffer == "True" || buffer == "1")
+    {
+        terminating = true;
+    }
+    else if (buffer == "false" || buffer == "False" || buffer == "0")
+    {
+        terminating = false;
+    }
+    else
+    {
+        SWSS_LOG_ERROR("Invalid terminating : %s", buffer.c_str());
+        throw std::invalid_argument("Invalid terminating");
+    }
+}
+
+template<>
+inline void lexical_convert(const string &buffer, vector<uint8_t> &protocols)
+{
+    SWSS_LOG_ENTER();
+
+    auto tokens = tokenize(buffer, ',');
+    protocols.clear();
+    protocols.reserve(tokens.size());
+    for (auto &token : tokens)
+    {
+        uint32_t protocol;
+        lexical_convert(token, protocol);
+        protocols.push_back(static_cast<uint8_t>(protocol));
+    }
+}
+
+template<>
+inline void lexical_convert(const string &buffer, vector<sai_ip_address_t> &ips)
+{
+    SWSS_LOG_ENTER();
+
+    auto tokens = tokenize(buffer, ',');
+    ips.clear();
+    ips.reserve(tokens.size());
+    for (auto &token : tokens)
+    {
+        sai_ip_address_t ip;
+        IpAddress ipAddress(token);
+        swss::copy(ip, ipAddress);
+        ips.push_back(ip);
+    }
+}
+
+template<>
+inline void lexical_convert(const string &buffer, std::vector<sai_u16_range_t> &ports)
+{
+    SWSS_LOG_ENTER();
+
+    auto tokens = tokenize(buffer, ',');
+    ports.clear();
+    ports.reserve(tokens.size());
+    for (auto &token : tokens)
+    {
+        sai_u16_range_t port;
+        if (token.find('-') == std::string::npos)
+        {
+            // Only one port
+            lexical_convert(token, port.min);
+            port.max = port.min;
+        }
+        else if (!extractVariables(token, '-', port.min, port.max))
+        {
+            // Range
+            SWSS_LOG_ERROR("Invalid port range : %s", token.c_str());
+            throw invalid_argument("Invalid port range");
+        }
+        ports.push_back(port);
+    }
+}
+
+}
+
+static bool operator==(const sai_ip_address_t& a, const sai_ip_address_t& b)
+{
+    SWSS_LOG_ENTER();
+
+    return a.addr_family == b.addr_family &&
+        (a.addr_family == SAI_IP_ADDR_FAMILY_IPV4 ? a.addr.ip4 == b.addr.ip4 : memcmp(a.addr.ip6, b.addr.ip6, sizeof(a.addr.ip6)) == 0);
+}
+
+static bool operator==(const sai_u16_range_t& a, const sai_u16_range_t& b)
+{
+    SWSS_LOG_ENTER();
+
+    return a.min == b.min && a.max == b.max;
+}
+
+template<class T>
+static bool updateValue(
+    const DashAclOrch::TaskArgs &ta,
+    const string &field,
+    boost::optional<T> &opt)
+{
+    SWSS_LOG_ENTER();
+
+    T value;
+
+    if (!getValue(ta, field, value))
+    {
+        return false;
+    }
+
+    if (opt && opt.value() == value)
+    {
+        return false;
+    }
+
+    opt = value;
+
+    return true;
+}
+
+sai_attr_id_t getSaiStage(DashAclDirection d, sai_ip_addr_family_t f, DashAclStage s)
+{
+    const static map<tuple<DashAclDirection, sai_ip_addr_family_t, DashAclStage>, sai_attr_id_t> StageMaps =
```

Comment:
Would it be better to declare this outside this function just so the code is a bit cleaner?

Reply:
IMHO, it's better to set the scope of a variable as small as possible.
Meanwhile, the initialization point of a static variable in a function will be determined at the first time that the function is called.  But if it's a global variable, the initialization sequence will be hard to determined. I believe a certain initialization sequence will be easier for maintaining if we want to expose these variables to other modules in the future.